src/Roslyn.Diagnostics.Analyzers/Core/Extensions/INamedTypeSymbolExtensions.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+
+namespace Roslyn.Diagnostics.Analyzers.Extensions
+{
+    internal static class INamedTypeSymbolExtensions
+    {
+        public static bool IsTestAttribute(this INamedTypeSymbol attributeClass, ConcurrentDictionary<INamedTypeSymbol, bool> knownTestAttributes, INamedTypeSymbol factAttribute)
+        {
+            if (knownTestAttributes.TryGetValue(attributeClass, out var isTest))
+                return isTest;
+
+            return knownTestAttributes.GetOrAdd(attributeClass, ExtendsFactAttribute(attributeClass, factAttribute));
+        }
+
+        private static bool ExtendsFactAttribute(INamedTypeSymbol namedType, INamedTypeSymbol factAttribute)
```

Comment:
You can use the `Inherits` extension method: https://github.com/dotnet/roslyn-analyzers/blob/f06e25d8ac47ae5a731107c7c82d9d09f343b8ba/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/MarkMembersAsStatic.cs#L199

Reply:
I used `DerivesFrom` and optimized it for the clear base class case.